### PR TITLE
Implement listen()

### PR DESCRIPTION
**Pull request by DK-DARKmatter**
Implement listen() function.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -282,12 +282,35 @@ pub extern "C" fn setsockopt(
...
+    if next!(listen(fd, n)) == -1 {
+        return -1;
+    }
+
+    *sock = match *sock {
+        Socket::Bound(p) => Socket::Listening(p),
+        _ => return error(libc::EBADFD, -1),
+    };
+
+    0
```
Comment (npmccallum) [inline src/lib.rs]: Don't use two match statements. Put lines 301-310 inside the match statement on line 297.

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 2]
```diff
@@ -282,12 +282,33 @@ pub extern "C" fn setsockopt(
...
+                return -1;
+            }
+
+            *sock = Socket::Listening(p);
+        },
+
+        _ => return error(libc::EBADFD, -1),
+    }
+
+    0
```
Comment (npmccallum) [inline src/lib.rs]: Move the `0` into line 302. Then you can remove the `return` keyword on line 305.

File: src/lib.rs [hunk 3]
```diff
@@ -282,12 +282,33 @@ pub extern "C" fn setsockopt(
...
+    let lock = match INDEX.get(fd) {
+        Some(s) => s,
+        None => return next!(listen(fd, n)),
+    };
 
-#[no_mangle]
+    let mut sock = lock.write().unwrap();
+
+    match *sock {
+        Socket::Bound(p) => {
```
Comment (npmccallum) [inline src/lib.rs]: I think the code below should look like this:

```rust
match next!(listen(fd, n)) {
  0 => { *sock = Socket::Listening(p); 0 },
  r => r,
}
```

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 4]
```diff
@@ -194,7 +194,7 @@ pub extern "C" fn setsockopt(
 ) -> c_int {
 }*/
 
-#[no_mangle]
+/*#[no_mangle]
```
Comment (frozencemetery) [inline src/lib.rs]: Why does this get removed?

File: src/lib.rs [hunk 5]
```diff
@@ -282,12 +282,33 @@ pub extern "C" fn setsockopt(
...
+pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {
+    let lock = match INDEX.get(fd) {
+        Some(s) => s,
+        None => return next!(listen(fd, n)),
+    };
 
-#[no_mangle]
+    let mut sock = lock.write().unwrap();
+
+    match *sock {
```
Comment (frozencemetery) [inline src/lib.rs]: Personally I prefer an `if let` for the outer match:

```rust
if let Socket::Bound(p) = *sock {
    match next!(listen(fd(n)) {
        /* ... */
    }
}
error(libc::EBADFD, -1)
```

Saves you a line and more importantly a layer of indentation.

**Review by frozencemetery — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 5 above)
Comment (npmccallum) [reply src/lib.rs]: Agreed.

**Comment by DK-DARKmatter** [comment]
Does this look better now? Can we merge this first?

**Review by frozencemetery — APPROVED** [review]

**Pull request merged**